allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -33,7 +33,32 @@ def index_with(self, vocab: Vocabulary):
         self.vocab = vocab
 
 
-class _LazyInstances(IterableDataset):
+class AllennlpLazyDataset(IterableDataset):
+    def __init__(
+        self,
+        deserialize: Callable[[str], Instance] = None,
+        serialize: Callable[[Instance], str] = None,
+        vocab: Vocabulary = None,
+    ) -> None:
+        super().__init__()
+        self.deserialize = deserialize
+        self.serialize = serialize
+        self.vocab = vocab
+
+    def index_with(self, vocab: Vocabulary):
+        self.vocab = vocab
+
+    def __len__(self):
+        """
+        We rely in a couple of places that calling len on the dataloader
+        (which in turn calls len on the dataset) doesn't raise an error.
+        In the case that you have an IterableDataset and you call len, the pytorch dataloader
+        actually spits out a warning - but we need actually calling it to not crash.
+        """
+        return 1
```

Comment:
I think we decided in another thread that this should crash, instead of returning 1, didn't we?  But you're just moving this logic, not changing it, so if you want to leave that for a separate PR, that's fine with me.

Reply:
Yes, that's what I was thinking. There is enough going on in this one as it is.